### PR TITLE
Apollo Federation Logging

### PR DESCRIPTION
**Pull request by TylerBloom**
This PR includes an initial pass in how logging will be approached in `apollo-federation`. There are two sets of changes:
 - Porting debug logs from JS
 - Added snapshot logging for visualization

Snapshot logs are feature gated (behind the new `snapshot_tracing` feature) as they contain a lot of information that is not desired in production and is largely designed for debugging. Also, the snapshot logging serializes objects, so deriving `Serialize` was added to many types.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-federation/src/operation/mod.rs [hunk 1]
```diff
@@ -603,24 +606,33 @@ pub(crate) use selection_map::SelectionValue;
...
 /// * directive arguments order does not matter (they get automatically sorted by their names).
 /// * selection cannot specify @defer directive
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
 pub(crate) enum SelectionKey {
     Field {
         /// The field alias (if specified) or field name in the resulting selection set.
         response_name: Name,
         /// directives applied on the field
+        // TODO(@TylerBloom): Add back in once apollo-compiler supports serde or a workaround is
```
Comment (dariuszkuc) [inline apollo-federation/src/operation/mod.rs:615]: can you open a follow up JIRA to track this?

Comment (goto-bus-stop) [reply apollo-federation/src/operation/mod.rs:615]: Suggest using a `serde(serialize_with)` that just uses `directives.to_string()`?

**Review by dariuszkuc — APPROVED** [review]

File: apollo-federation/src/operation/mod.rs (hunk 1 above)
Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:615]: > Suggest using a `serde(serialize_with)` that just uses `directives.to_string()`?

That's a good idea. I'll add replace the relevant `serde(skip)`s will that.

Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:615]: Here's the Jira ticket for the other `TODO`s.
https://apollographql.atlassian.net/browse/FED-363

File: apollo-federation/src/operation/mod.rs [hunk 2]
```diff
@@ -66,7 +67,7 @@ static NEXT_ID: atomic::AtomicUsize = atomic::AtomicUsize::new(1);
 ///
 /// Note that we shouldn't add `derive(Serialize, Deserialize)` to this without changing the types
 /// to be something like UUIDs.
-#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize)]
```
Comment (abernix) [inline apollo-federation/src/operation/mod.rs:70]: @TylerBloom @lennyburdette @dariuszkuc Does the comment above this need to be updated after this change?  (The comment says not to add `Serialize` here, but adds `Serialize`? 😄

Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:70]: Ah, I definitely should have made a comment about this during review. That's my bad. My thought was that serialize was ok but deserialize was not (and both was definitely not a good idea); however, this should probably be feature-gated like the `snapshot!` macro that was added. I need to do some clean up around the `Serialize` implementation of a few types. I can bundle this change into that PR.

Comment (abernix) [reply apollo-federation/src/operation/mod.rs:70]: Thank you! that'd be wonderful. 🙇  Consider my curiousity placated. 😄

Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:70]: Addressed in this PR: https://github.com/apollographql/router/pull/5724

**Pull request merged**